appcues/src/main/java/com/appcues/RenderContextManager.kt

```diff
@@ -1,25 +1,80 @@
 package com.appcues
 
+import com.appcues.analytics.ExperienceLifecycleTracker
+import com.appcues.data.model.Experience
+import com.appcues.data.model.ExperienceTrigger
+import com.appcues.data.model.ExperienceTrigger.Qualification
 import com.appcues.data.model.RenderContext
 import com.appcues.data.model.RenderContext.Embed
+import com.appcues.statemachine.StateMachine
+import org.koin.core.component.KoinScopeComponent
+import org.koin.core.component.get
+import org.koin.core.component.inject
+import org.koin.core.scope.Scope
 import java.lang.ref.WeakReference
 
-internal class RenderContextManager {
+internal class RenderContextManager(override val scope: Scope) : KoinScopeComponent {
 
-    private val slots: HashMap<RenderContext, WeakReference<AppcuesFrameView>> = hashMapOf()
+    private val lifecycleTracker by inject<ExperienceLifecycleTracker>()
+
+    private val frameSlots: HashMap<RenderContext, WeakReference<AppcuesFrameView>> = hashMapOf()
+
+    private val stateMachineSlots: HashMap<RenderContext, StateMachine> = hashMapOf()
+
+    private val potentiallyRenderableExperiences = hashMapOf<RenderContext, List<Experience>>()
 
     fun registerEmbedFrame(frameId: String, frame: AppcuesFrameView) {
-        slots[Embed(frameId)] = WeakReference(frame)
+        val renderContext = Embed(frameId)
+
+        // silently removing the existing state machine to ensure we get a new one when trying
+        // to start an experience for this render context
+        stateMachineSlots.remove(renderContext)
+        frameSlots[Embed(frameId)] = WeakReference(frame)
     }
 
     fun getEmbedFrame(renderContext: RenderContext): AppcuesFrameView? {
         return when (renderContext) {
-            is Embed -> slots[renderContext]?.get()
+            is Embed -> frameSlots[renderContext]?.get()
             else -> null
         }
     }
 
-    fun clear() {
-        slots.clear()
+    fun getOrCreateStateMachines(renderContext: RenderContext): StateMachine {
+        return stateMachineSlots[renderContext] ?: run {
+            get<StateMachine>()
+                .also { stateMachineSlots[renderContext] = it }
+                .also { lifecycleTracker.start(it, { potentiallyRenderableExperiences.remove(renderContext) }) }
+        }
+    }
+
+    fun getStateMachine(renderContext: RenderContext): StateMachine? {
+        return stateMachineSlots[renderContext]
+    }
+
+    fun putExperiences(experiences: List<Experience>, trigger: ExperienceTrigger) {
```

Comment:
just moving logic around, here we have the new experience list and the trigger, then we check if its screen_view in order to run some cleanup